File: sdk/ts/src/validators/tls.ts
import { ValidationResult } from "../types"

export function validateTls(url: string): ValidationResult {
  try {
    const parsed = new URL(url)
    if (parsed.protocol !== "https:") {
      return {
        valid: false,
        reason: `[DoorNo.402] BLOCKED -- payment refused over non-TLS connection: ${url}`,
      }
    }
    return { valid: true, reason: "TLS ok" }
  } catch {
    return { valid: false, reason: "invalid URL" }
  }
}

export function validateRedirect(
  originalUrl: string,
  finalUrl: string
): ValidationResult {
  try {
    const orig = new URL(originalUrl)
    const final_ = new URL(finalUrl)
    if (orig.hostname !== final_.hostname) {
      return {
        valid: false,
        reason: `[DoorNo.402] BLOCKED -- redirect hijack: ${orig.hostname} -> ${final_.hostname}`,
      }
    }
    if (orig.protocol === "https:" && final_.protocol === "http:") {
      return {
        valid: false,
        reason: "[DoorNo.402] BLOCKED -- TLS downgrade via redirect",
      }
    }
    return { valid: true, reason: "redirect is safe" }
  } catch {
    return { valid: false, reason: "invalid URL" }
  }
}

File: sdk/python/doorno402/validators/tls.py
"""VULN-06: TLS Enforcer"""
from urllib.parse import urlparse


def validate_tls(url: str) -> dict:
    parsed = urlparse(url)
    if parsed.scheme != "https":
        # Exempt localhost for testing, EXCEPT port 3005 (which is the VULN-06 TLS demo)
        if parsed.hostname in ["localhost", "127.0.0.1"] and parsed.port != 3005:
            return {"valid": True, "reason": "localhost exempt from TLS"}
            
        return {
            "valid": False,
            "reason": f"[DoorNo.402] BLOCKED -- payment refused over "
                      f"non-TLS connection: {url}",
        }
    return {"valid": True, "reason": "TLS ok"}
